### PR TITLE
[ZEPPELIN-6204] do not allow init params in JDBC URLs for H2

### PR DESCRIPTION
**Pull request by pjfanning**
### What is this PR for?
ZEPPELIN-6204

Slight tidy of the existing disallow list for strings in JDBC urls so that they are checked against just the query params and not the hostname in the URL. 

### What type of PR is it?
Bug Fix

### Todos
* [ ] - Task

### What is the Jira issue?
* Open an issue on Jira https://issues.apache.org/jira/browse/ZEPPELIN/
* Put link here, and add [ZEPPELIN-*Jira number*] in PR title, eg. [ZEPPELIN-533]

### How should this be tested?
* Strongly recommended: add automated unit tests for any new or changed behavior
* Outline any manual steps to test the PR here.

### Questions:
* Does the license files need to update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by Reamer** [comment]
I don't know whether we are preventing features that others want to use with these changes. 

Doesn't it make more sense to secure the H2 server according to the requirements? After all, Zeppelin is just one JDBC client of many.

**Comment by pjfanning** [comment]
@Reamer @jongyoul is it ok to consider this for merge now? There are probably a few more query params that could be added to the disallow list but this covers some of the more commonly mentioned ones.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 1]
```diff
@@ -589,17 +597,123 @@ public Connection getConnection(InterpreterContext context)
...
+
+    if (containsKeyIgnoreCase(params, ALLOW_LOAD_LOCAL) ||
+            containsKeyIgnoreCase(params, ALLOW_LOAD_LOCAL_IN_FILE_NAME) ||
+            containsKeyIgnoreCase(params, ALLOW_LOCAL_IN_FILE_NAME) ||
+            containsKeyIgnoreCase(params, ALLOW_URL_IN_LOCAL_IN_FILE_NAME) ||
+            containsKeyIgnoreCase(params, ALLOW_LOAD_LOCAL_IN_FILE_IN_PATH) ||
+            containsKeyIgnoreCase(params, AUTO_DESERIALIZE) ||
+            containsKeyIgnoreCase(params, SOCKET_FACTORY)) {
+      throw new IllegalArgumentException("Connection URL contains sensitive configuration");
+    }
```
Comment (seung-00) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:612]:
```suggestion
       Stream.of(ALLOW_LOAD_LOCAL, ALLOW_LOCAL_IN_FILE_NAME, ...).forEach(rule -> {
            if (containsKeyIgnoreCase(params, rule)) throw new IllegalArgumentException("Connection URL contains sensitive configuration: " + rule);
        });
```

I find this version easier to read.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 2]
```diff
@@ -589,17 +597,123 @@ public Connection getConnection(InterpreterContext context)
   }
 
   private void validateConnectionUrl(String url) {
-    String decodedUrl;
-    decodedUrl = URLDecoder.decode(url, StandardCharsets.UTF_8);
+    final String decodedUrl = urlDecode(url, url, 0);
+    final Map<String, String> params = parseUrlParameters(decodedUrl);
```
Comment (tbonelee) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:601]: Just a small thought. I was wondering if decoding the full URL before parsing could lead to unexpected behavior in some edge cases. For example, if a parameter value contains encoded characters like `%26` (&) or `%3D` (=), decoding them too early might interfere with how the parameters are parsed.

I’m not sure if values like passwords are actually allowed or commonly used in this context, so this might not be relevant. But I thought it was worth mentioning just in case. 🙏

**Comment by tbonelee** [comment]
Just a small thought. I was wondering if decoding the full URL before parsing could lead to unexpected behavior in some edge cases. For example, if a parameter value contains encoded characters like `%26` (&) or `%3D` (=), decoding them too early might interfere with how the parameters are parsed.

I’m not sure if values like passwords are actually allowed or commonly used in this context, so this might not be relevant. But I thought it was worth mentioning just in case. 🙏

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 3]
```diff
@@ -589,17 +597,123 @@ public Connection getConnection(InterpreterContext context)
...
 
+  /**
+   * Decode the URL encoded string recursively until no more decoding is needed.
+   * This is to handle cases where the URL might be double-encoded.
+   *
+   * @param url the original URL (for logging purposes)
+   * @param encoded the URL encoded string
+   * @param recurseCount the current recursion depth
+   * @return the decoded string
+   * @throws IllegalArgumentException if the recursion depth exceeds 100
```
Comment (seung-00) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: The function seems to recurse up to 10 times.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 4]
```diff
@@ -156,14 +156,22 @@ public class JDBCInterpreter extends KerberosInterpreter {
...
   private static final String ALLOW_LOCAL_IN_FILE_NAME = "allowLocalInfile";
 
   private static final String ALLOW_URL_IN_LOCAL_IN_FILE_NAME = "allowUrlInLocalInfile";
 
+  private static final String AUTO_DESERIALIZE = "autoDeserialize";
+
+  private static final String SOCKET_FACTORY = "socketFactory";
+
+  private static final String INIT = "INIT";
+
```
Comment (seung-00) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:174]: could be nice to group these into a Set.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 5]
```diff
@@ -589,17 +597,123 @@ public Connection getConnection(InterpreterContext context)
...
+    // MySQL supports parentheses in the URL
+    // https://dev.mysql.com/doc/connectors/en/connector-j-reference-jdbc-url-format.html
+    // eg jdbc:mysql://(host=myhost,port=1111,allowLoadLocalInfile=true)/db
+    int parensIndex = extractFromParens(url, 0, parameters);
+    while (parensIndex > 0) {
+      parensIndex = extractFromParens(url, parensIndex, parameters);
+    }
+
+    // Split the URL into the base part and the parameters part
+    String[] parts = url.split("[?&;]");
```
Comment (seung-00) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:655]: could be a bit faster if the regex was cached to avoid rebuilding every time

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java [hunk 6]
```diff
@@ -589,17 +597,123 @@ public Connection getConnection(InterpreterContext context)
...
+      return -1;
+    }
+    final int endIndex = input.indexOf(')', startIndex);
+    if (startIndex != -1 && endIndex != -1) {
+      String params = input.substring(startIndex + 1, endIndex);
+      String[] keyValuePairs = params.split(",");
+      for (String pair : keyValuePairs) {
+        splitNameValue(pair, parameters);
+      }
+    }
```
Comment (seung-00) [inline jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:697]: Does it work well if parentheses are in the URL query parameters?
`jdbc:mysql://.../db?password=abc(def)`

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 1 above)
Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:612]: I can't commit this unless it compiles so please don't use the suggestion option to suggest incomplete code.
I also disagree that this is more readable. The existing code in this PR is closer to the existing style in the existing codebase.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 4 above)
Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:174]: I'm happy to stick with the existing code style.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 3 above)
Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java]: thanks - I fixed the comment. I recently reduced the recurse count from 100

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 5 above)
Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:655]: the cost of this regex will be tiny compared to the I/O cost of using JDBC commands so I'm not very interested in optimising every line of code here

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 6 above)
Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:697]: the code doesn't care about what the values are, it only cares about the param names - write some test cases if you are suspicious

**Comment by pjfanning** [comment]
> Just a small thought. I was wondering if decoding the full URL before parsing could lead to unexpected behavior in some edge cases. For example, if a parameter value contains encoded characters like `%26` (&) or `%3D` (=), decoding them too early might interfere with how the parameters are parsed.
> 
> I’m not sure if values like passwords are actually allowed or commonly used in this context, so this might not be relevant. But I thought it was worth mentioning just in case. 🙏

The code in this PR ignores the param values. It only worries about the param names.
The original URL as entered by the user is passed to the JDBC layer - my code splits up the URL but only for my checks for suspicious param names.

File: jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java (hunk 6 above)
Comment (seung-00) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:697]: I tested the following cases:
```java
testBannedMySQLQueryParamWithValidParam("password=allowLoadLocalInfile"); // 1. fail
testBannedMySQLQueryParamWithValidParam("password=(allowLoadLocalInfile)"); // 2. pass
```
I might be missing something, but test case 2 seems unintended. 
Part of the value `allowLoadLocalInfile` appears to be treated as a key due to the parentheses.

Comment (pjfanning) [reply jdbc/src/main/java/org/apache/zeppelin/jdbc/JDBCInterpreter.java:697]: thanks - I'll look into the issue - but in the end of the day, it's not a very likely password
and the existing checks before this PR, they just do a string match on the whole URL so that code wouldn't handle this either

**Comment by tbonelee** [comment]
@pjfanning Thanks for the clarification. I took a closer look and you're right. Decoding encoded special characters in values before validation would at most introduce an extra key, but it does not interfere with the parsing of subsequent parameters. I had initially worried about false negatives (i.e. cases where disallowed parameters might go undetected), but that doesn't seem to be the case. Thanks again, and sorry for the confusion.

**Comment by pjfanning** [comment]
@jongyoul @Reamer would you be able to review this change?

**Comment by pjfanning** [comment]
@jongyoul @Reamer would you be able to review this change? I think this fixes a couple of issues in the existing implementation.

**Comment by pjfanning** [comment]
@jongyoul do you think this could be merged?

**Comment by pjfanning** [comment]
@jongyoul @Reamer I am not a committer so can't merge this. Would either of you be in a position to do so?

**Comment by tbonelee** [comment]
I’ve merged this into master.
Since I considered it a breaking change, I didn’t include it in branch-0.12.
Do you think it needs to be backported?

**Comment by pjfanning** [comment]
Thanks @tbonelee - I would regard it more as a bug fix or fixes. It shouldn't change the behaviour for any use cases that we want to support.
